language-support/hs/bindings/test/DA/Ledger/Sandbox.hs

```diff
@@ -0,0 +1,123 @@
+
+data Sandbox = Sandbox { port :: Port, proh :: ProcessHandle }
+
+selectedPort :: Int
+selectedPort = 0 --dynamic port selection
+
+sandboxProcess :: SandboxSpec -> CreateProcess
+sandboxProcess SandboxSpec{dar} =
+    proc binary [ dar, "--port", show selectedPort]
+    where
+        binary = "ledger/sandbox/sandbox-binary"
+
+startSandboxProcess :: SandboxSpec -> IO (ProcessHandle,Maybe Handle)
+startSandboxProcess spec = do
+    (_,hOutOpt,_,proh) <-
+        createProcess (sandboxProcess spec) {
+        std_out = CreatePipe,
+        std_err = CreatePipe, -- Question: ought the pipe to be drained?
+        create_group = True  -- To avoid sending INT to ourself
+        }
+    pid <- getPid proh
+    log $ "Sandbox process started, pid = " <> show pid
+    return (proh,hOutOpt)
+
+shutdownSandboxProcess :: ProcessHandle -> IO ()
+shutdownSandboxProcess proh = do
+    pidOpt <- getPid proh
+    log $ "Sending INT to sandbox process: " <> show pidOpt
+    interruptProcessGroupOf proh
+    x <- timeout 5 "Sandbox process didn't exit" (waitForProcess proh)
+    log $ "Sandbox process exited with: " <> show x
+    return ()
+
+parsePortFromListeningLine :: String -> IO Port
+parsePortFromListeningLine line = do
+    [_,portNumStr] <- return (splitOn ":" line)
+    num <- evaluate (read portNumStr)
+    return (Port num)
+
+interestingLineFromSandbox :: String -> Bool
+interestingLineFromSandbox line =
+    any (`isInfixOf` line)
+    [--"listening",
+     "error", "Address already in use", "java.net.BindException"]
+
+getListeningLine :: Handle -> IO String
+getListeningLine h = loop where
+    loop = do
+        line <- hGetLine h
+        when (interestingLineFromSandbox line) $ log $ "SANDBOX: " <> line
+        if "listening" `isInfixOf` line
+            then return line
+            else if "initialization error" `isInfixOf` line
+                 then error line
+                 else loop
+
+discoverListeningPort :: Maybe Handle -> IO Port
+discoverListeningPort hOpt = do
+    Just h <- return hOpt
+    log "Looking for sandbox listening port..."
+    line <- getListeningLine h
+    port <- parsePortFromListeningLine line
+        `onException` log ("Failed to parse listening port from: " <> show line)
+    log $ "Sandbox listening on port: " <> show (unPort port)
+    return port
+
+startSandbox :: SandboxSpec-> IO Sandbox
+startSandbox spec = do
+    (proh,hOpt) <-startSandboxProcess spec
+    port <-
+        timeout 5 "Didn't discover sandbox port" (discoverListeningPort hOpt)
+        `onException` shutdownSandboxProcess proh
+    return Sandbox { port, proh }
+
+shutdownSandbox :: Sandbox -> IO ()
+shutdownSandbox Sandbox{proh} = do shutdownSandboxProcess proh
+
+withSandbox :: SandboxSpec -> (Sandbox -> IO a) -> IO a
+withSandbox spec f =
+    bracket (startSandbox spec)
+    shutdownSandbox
+    f
+
+data Timeout = Timeout deriving Show
+instance Exception Timeout
+
+timeout :: Seconds -> String -> IO a -> IO a
```

Comment:
Why not just use `System.Time.Extra.timeout`? Inventing your own timeout handler is not a good plan.

Reply:
Changed to use `System.Time.Extra.timeout` as suggested.
But this only replaces the use of `race`. The rest of the wrapper (now named `timeoutError`) persists.